servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapMerge.java

```diff
@@ -0,0 +1,681 @@
+                terminalNotificationUpdater = newUpdater(FlatMapSubscriber.class, TerminalNotification.class,
+                "terminalNotification");
+
+        @Nullable
+        private volatile CompositeException delayedError;
+        @SuppressWarnings("unused")
+        private volatile int tryRequestMoreLock;
+        @SuppressWarnings("unused")
+        private volatile int emittingLock;
+        private volatile long pendingDemand;
+        private volatile int activeMappedSources;
+        @SuppressWarnings("unused")
+        @Nullable
+        private volatile TerminalNotification terminalNotification;
+
+        /**
+         * This variable is only accessed within the "emitting lock" so we rely upon this to provide visibility to
+         * other threads.
+         */
+        private boolean targetTerminated;
+        @Nullable
+        private Subscription subscription;
+
+        private final Subscriber<? super R> target;
+        private final Queue<Object> pending;
+        private final PublisherFlatMapMerge<T, R> source;
+        private final Set<FlatMapPublisherSubscriber<T, R>> subscribers;
+
+        FlatMapSubscriber(PublisherFlatMapMerge<T, R> source, Subscriber<? super R> target) {
+            this.source = source;
+            this.target = target;
+            // Start with a small capacity as maxConcurrency can be large.
+            pending = newUnboundedMpscQueue(min(2, source.maxConcurrency));
+            subscribers = newSetFromMap(new ConcurrentHashMap<>());
+        }
+
+        @Override
+        public void cancel() {
+            doCancel(true);
+        }
+
+        @Override
+        public void request(final long n) {
+            assert subscription != null;
+            if (!isRequestNValid(n)) {
+                subscription.request(n);
+            } else {
+                pendingDemandUpdater.accumulateAndGet(this, n, FlowControlUtils::addWithOverflowProtection);
+                tryRequestMore();
+            }
+        }
+
+        @Override
+        public void onSubscribe(final Subscription s) {
+            if (!checkDuplicateSubscription(subscription, s)) {
+                return;
+            }
+
+            // We assume that FlatMapSubscriber#cancel() will never be called before this method, and therefore we
+            // don't have to worry about being cancelled before the onSubscribe method is called.
+            subscription = ConcurrentSubscription.wrap(s);
+            target.onSubscribe(this);
+
+            // Currently we always request maxConcurrency elements from upstream.
+            subscription.request(source.maxConcurrency);
+        }
+
+        @Override
+        public void onNext(@Nullable final T t) {
+            final Publisher<? extends R> publisher = requireNonNull(source.mapper.apply(t));
+            FlatMapPublisherSubscriber<T, R> subscriber = new FlatMapPublisherSubscriber<>(this);
+            if (subscribers.add(subscriber)) {
```

Comment:
how could this ever return false ? `FlatMapPublisherSubscriber` doesn't have an equals implementation and so this will always return true imho.

Reply:
it isn't expected to fail ... but if it did we would be in an inconsistent state if we attempted to modify the other state.

Reply:
I'll add a throw in an `else` statement to clarify.

Reply:
I still don't understand how this ever could return `false` even in an "inconsistent state".

Reply:
as mentioned above it isn't expected to fail. is your suggestion to ignore the return value?

Reply:
I'll just change it to an assert and add a comment. let me know if you had something else in mind.